### PR TITLE
add support for mutation webhooks in Istio's newer versions

### PR DESCRIPTION
**Pull request by piotrkpc**
See also: https://github.com/kyma-incubator/reconciler/issues/456

### CODE REVIEW

File: pkg/reconciler/instances/istio/actions/performer.go [hunk 1]
```diff
@@ -219,6 +233,19 @@ func (c *DefaultIstioPerformer) PatchMutatingWebhook(context context.Context, ku
 	return nil
 }
 
+func (c *DefaultIstioPerformer) selectWebhookToPatch(context context.Context, candidatesNames []string, clientSet clientgo.Interface) (*v1.MutatingWebhookConfiguration, error) {
```
Comment (mjakobczyk) [inline pkg/reconciler/instances/istio/actions/performer.go:236]: Please name return parameters here and re-use them in the function body

File: pkg/reconciler/instances/istio/actions/performer.go [hunk 2]
```diff
@@ -219,6 +233,19 @@ func (c *DefaultIstioPerformer) PatchMutatingWebhook(context context.Context, ku
...
+	var wh *v1.MutatingWebhookConfiguration
+	var err error
```
Comment (mjakobczyk) [inline pkg/reconciler/instances/istio/actions/performer.go:238]: Please move this var to named return parameters

File: pkg/reconciler/instances/istio/actions/performer.go (hunk 1 above)
Comment (mjakobczyk) [inline pkg/reconciler/instances/istio/actions/performer.go:236]: I would rather call this function: `selectFirstMatchingWebhookFromCandidates` - we don't need to mention patching here

File: pkg/reconciler/instances/istio/actions/performer.go [hunk 3]
```diff
@@ -190,6 +191,20 @@ func (c *DefaultIstioPerformer) Install(kubeConfig, istioChart, version string,
 }
 
 func (c *DefaultIstioPerformer) PatchMutatingWebhook(context context.Context, kubeClient kubernetes.Client, logger *zap.SugaredLogger) error {
+	clientSet, err := kubeClient.Clientset()
+	if err != nil {
+		return err
+	}
+
+	primary := "istio-revision-tag-default"
```
Comment (mjakobczyk) [inline pkg/reconciler/instances/istio/actions/performer.go:199]: Please move this value to a dedicated const

File: pkg/reconciler/instances/istio/actions/performer.go [hunk 4]
```diff
@@ -190,6 +191,20 @@ func (c *DefaultIstioPerformer) Install(kubeConfig, istioChart, version string,
...
+	secondary := "istio-sidecar-injector"
```
Comment (mjakobczyk) [inline pkg/reconciler/instances/istio/actions/performer.go:200]: Please move this value to a dedicated const

**Review by mjakobczyk — CHANGES_REQUESTED** [review]

File: pkg/reconciler/instances/istio/actions/performer.go [hunk 5]
```diff
@@ -219,6 +233,19 @@ func (c *DefaultIstioPerformer) PatchMutatingWebhook(context context.Context, ku
...
 
+func (c *DefaultIstioPerformer) selectWebhookToPatch(context context.Context, candidatesNames []string, clientSet clientgo.Interface) (*v1.MutatingWebhookConfiguration, error) {
+	var wh *v1.MutatingWebhookConfiguration
+	var err error
+	for _, webhookName := range candidatesNames {
+		wh, err = clientSet.AdmissionregistrationV1().MutatingWebhookConfigurations().Get(context, webhookName, metav1.GetOptions{})
+		if err != nil {
+			continue
+		}
+		return wh, err
```
Comment (mjakobczyk) [inline pkg/reconciler/instances/istio/actions/performer.go:244]: After you give names to return parameters you can simply `return` from the function and they will be returned automatically

**Review by mjakobczyk — CHANGES_REQUESTED** [review]

**Review by mjakobczyk — APPROVED** [review]

File: pkg/reconciler/instances/istio/actions/performer.go [hunk 6]
```diff
@@ -190,26 +179,46 @@ func (c *DefaultIstioPerformer) Install(kubeConfig, istioChart, version string,
...
+		Values:   []string{"kube-system"},
+	}
+
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		whConf, err := c.selectWebhookConfFormCandidates(context, candidatesNames, clientSet)
+		if err != nil {
+			return err
+		}
+		var whFound bool
+		for i := range whConf.Webhooks {
```
Comment (mjakobczyk) [inline pkg/reconciler/instances/istio/actions/performer.go:205]: You can wrap this logic to a private function in the same way as with `c.selectWebhookConfFormCandidates`

File: pkg/reconciler/instances/istio/actions/performer.go [hunk 7]
```diff
@@ -190,26 +179,46 @@ func (c *DefaultIstioPerformer) Install(kubeConfig, istioChart, version string,
...
+				whConf.Webhooks[i].NamespaceSelector.MatchExpressions = append(
+					whConf.Webhooks[i].NamespaceSelector.MatchExpressions,
+					requiredLabelSelector,
+				)
+			}
+		}
+		if !whFound {
+			return fmt.Errorf("could not find webhook %s in WebhookConfiguration %s", webhookNameToChange, whConf.Name)
+		}
+		_, err = clientSet.AdmissionregistrationV1().
```
Comment (mjakobczyk) [inline pkg/reconciler/instances/istio/actions/performer.go:217]: IMO we should align code formatting - either like this or like with other one-liners such as: `clientSet.AdmissionregistrationV1().MutatingWebhookConfigurations().Get(context, webhookName, metav1.GetOptions{})`

File: pkg/reconciler/instances/istio/actions/performer.go (hunk 6 above)
Comment (piotrkpc) [reply pkg/reconciler/instances/istio/actions/performer.go]: done

**Review by mjakobczyk — APPROVED** [review]

**Review by werdes72 — APPROVED** [review]
/meow

**Pull request merged**